File: sample_question/custom_validator/sample-validator.php
<?php

// Sample interactive validator for the "guess the secret number" problem.
//
// Reads three secret values from stdin (max_value, max_tries, secret), reports the
// judge-supplied limits on stderr, then plays the guessing game with the contestant
// program. The exit code is the verdict.

const AC = 0;
const WA = 1;
const TLE = 2;
const RE = 3;
const PE = 4;

function finish(int $exitCode, string $message = ""): never
{
    if ($message !== "") {
        fwrite(STDERR, $message . "\n");
    }
    exit($exitCode);
}

function readTrimmedLine(): ?string
{
    $line = fgets(STDIN);
    if ($line === false) {
        return null;
    }
    return rtrim($line, "\r\n \t");
}

function parseLong(string $text): ?int
{
    if (preg_match('/^[+-]?\d+$/', $text) !== 1) {
        return null;
    }
    return (int) $text;
}

function envNumber(string $name): string
{
    $value = getenv($name);
    return $value === false ? "null" : $value;
}

function printLimits(): void
{
    $userLanguage = getenv("USER_LANGUAGE");
    $perLanguageLimits = getenv("PER_LANGUAGE_LIMITS");

    fwrite(
        STDERR,
        "{"
            . '"problem_time_limit": ' . envNumber("PROBLEM_TIME_LIMIT") . ", "
            . '"problem_output_limit": ' . envNumber("PROBLEM_OUTPUT_LIMIT") . ", "
            . '"problem_memory_limit": ' . envNumber("PROBLEM_MEMORY_LIMIT") . ", "
            . '"problem_pid_limit": ' . envNumber("PROBLEM_PID_LIMIT") . ", "
            . '"user_language": ' . ($userLanguage !== false ? '"' . $userLanguage . '"' : "null") . ", "
            . '"per_language_limits": ' . ($perLanguageLimits !== false ? $perLanguageLimits : "null")
            . "}\n"
    );
}

function readSecretValue(): int
{
    $line = readTrimmedLine();
    if ($line === null) {
        finish(RE, "Invalid secret data");
    }
    $value = parseLong($line);
    if ($value === null) {
        finish(RE, "Invalid secret data");
    }
    return $value;
}

$maxValue = readSecretValue();
$maxTries = readSecretValue();
$secret = readSecretValue();

printLimits();

echo $maxValue, "\n";
flush();

for ($iteration = 0; ; $iteration++) {
    $line = readTrimmedLine();
    if ($line === null) {
        break;
    }

    if ($line !== "" && $line[0] === "!") {
        $answer = parseLong(substr($line, 1));
        if ($answer === null) {
            finish(RE, "Invalid data");
        }
        if ($answer === $secret) {
            finish(AC);
        }
        finish(WA, "!" . $secret);
    }

    if ($iteration > $maxTries) {
        finish(TLE, "Exceeded number of allowed iterations");
    }

    $value = parseLong($line);
    if ($value === null) {
        finish(RE, "Invalid data");
    }

    echo ($secret < $value ? "<" : ">="), "\n";
    flush();
}

finish(RE, "Unexpected flow");

File: sample_question/custom_validator/sample-validator.scala
// Sample interactive validator for the "guess the secret number" problem.
//
// Reads three secret values from stdin (max_value, max_tries, secret), reports the
// judge-supplied limits on stderr, then plays the guessing game with the contestant
// program. The exit code is the verdict.

object Main {

  private val AC = 0
  private val WA = 1
  private val TLE = 2
  private val RE = 3
  private val PE = 4

  private def finish(exitCode: Int, message: String = ""): Nothing = {
    if (message.nonEmpty) Console.err.println(message)
    sys.exit(exitCode)
  }

  private def readTrimmedLine(): Option[String] = {
    val line = scala.io.StdIn.readLine()
    if (line == null) None
    else {
      var end = line.length
      while (end > 0 && (line.charAt(end - 1) == '\r' || line.charAt(end - 1) == ' ' ||
             line.charAt(end - 1) == '\t')) {
        end -= 1
      }
      Some(line.substring(0, end))
    }
  }

  private def parseLong(text: String): Option[Long] =
    if (text.matches("[+-]?\\d+")) text.toLongOption else None

  private def envNumber(name: String): String =
    Option(System.getenv(name)).getOrElse("null")

  private def printLimits(): Unit = {
    val userLanguage =
      Option(System.getenv("USER_LANGUAGE")).map(value => "\"" + value + "\"").getOrElse("null")
    val perLanguageLimits =
      Option(System.getenv("PER_LANGUAGE_LIMITS")).getOrElse("null")

    Console.err.println(
      "{" +
        "\"problem_time_limit\": " + envNumber("PROBLEM_TIME_LIMIT") + ", " +
        "\"problem_output_limit\": " + envNumber("PROBLEM_OUTPUT_LIMIT") + ", " +
        "\"problem_memory_limit\": " + envNumber("PROBLEM_MEMORY_LIMIT") + ", " +
        "\"problem_pid_limit\": " + envNumber("PROBLEM_PID_LIMIT") + ", " +
        "\"user_language\": " + userLanguage + ", " +
        "\"per_language_limits\": " + perLanguageLimits +
        "}"
    )
  }

  private def readSecretValue(): Long =
    readTrimmedLine().flatMap(parseLong).getOrElse(finish(RE, "Invalid secret data"))

  def main(args: Array[String]): Unit = {
    val maxValue = readSecretValue()
    val maxTries = readSecretValue()
    val secret = readSecretValue()

    printLimits()

    println(maxValue)
    Console.flush()

    var iteration = 0L
    var running = true

    while (running) {
      readTrimmedLine() match {
        case None => running = false
        case Some(line) =>
          if (line.nonEmpty && line.charAt(0) == '!') {
            parseLong(line.substring(1)) match {
              case None => finish(RE, "Invalid data")
              case Some(answer) =>
                if (answer == secret) finish(AC)
                finish(WA, "!" + secret)
            }
          } else {
            if (iteration > maxTries) finish(TLE, "Exceeded number of allowed iterations")
            parseLong(line) match {
              case None => finish(RE, "Invalid data")
              case Some(value) =>
                println(if (secret < value) "<" else ">=")
                Console.flush()
                iteration += 1
            }
          }
      }
    }

    finish(RE, "Unexpected flow")
  }
}
